### PR TITLE
HARMONY-1315: Fairer queueing and redesign of polling for work and getting service metrics

### PR DESCRIPTION
**Pull request by chris-durbin**
## Jira Issue ID
HARMONY-1315

## Description
When polling for work we now look at the currently in progress work to determine the next user that should have work performed.

We also introduced the user_work table in this pull request to significantly improve performance of polling for work and capturing the number of items currently in work.

When there were hundreds of thousands of work items for a service the prior polling for work queries could take more than 90 seconds and now they take a few hundred milliseconds with the same amount of work items. Similarly the query to determine the number of work items ready or running for a service used by the HPA now takes milliseconds instead of tens of seconds with many work items.

## Local Test Steps
There were no changes to user facing functionality other than the fair queueing algorithm change. For testing we should regression test certain areas and verify the database migration.

1. Submit a few requests that will be in progress at the time of running the database migration. Pause a couple of the requests.
2. Run the database migration and verify the contents of the user_work table make sense. Verify any non-paused jobs work to completion after the migration and restarting harmony. Verify the paused jobs do not have work performed.
3. Resume one of the paused jobs and make sure it is worked to completion.
4. Cancel one of the jobs and make sure that the entries for that job are removed from the user_work table and there are no problems with polling.

Note that I also tested this branch in a sandbox environment.

## PR Acceptance Checklist
* [X] Acceptance criteria met
* [X] Tests added/updated (if needed) and passing
* [X] Documentation updated (if needed)

### CODE REVIEW

File: app/models/user-work.ts [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+ */
+export async function deleteUserWorkForJob(tx: Transaction, jobID: string): Promise<number> {
+  const numDeleted = await tx(UserWork.table)
+    .where({ job_id: jobID })
+    .del();
+  return numDeleted;
+}
+
+/**
+ * Deletes all of the rows for the given job from the user_work table.
```
Comment (indiejames) [inline app/models/user-work.ts:124]: Looks like you copied this comment from the function above. It does not mention `serviceID`.

File: docs/fair-queueing.md [hunk 2]
```diff
@@ -25,18 +25,19 @@ Workers for each service are run as pods in a Kubernetes cluster. These workers
...
 ## The Algorithm
 The steps to perform fair queueing in harmony can be summarized as follows:
 
 When a worker requests a work item for a given service
 
 1. Identify all the users actively requesting work to be done on the service
-2. From the list of identified users, determine which user had work performed (for any service) least recently
-3. From that user's list of jobs (only those invoking the service) sort the jobs by the `isAsync` column to favor synchronous jobs, then sort by the `updatedAt` column to identify the job that has been worked least recently - return a work item for that job for the given service
-4. Update the work item and job timestamps in the database to indicate work being done at the current point in time
+2. From the list of identified users, determine which user has the least work currently in progress
```
Comment (indiejames) [inline docs/fair-queueing.md:37]: maybe change to "which user has the least total work (for all services) currently in progress"

**Review by indiejames — APPROVED** [review]
Just nits

File: app/models/user-work.ts (hunk 1 above)
Comment (chris-durbin) [reply app/models/user-work.ts]: Fixed

File: app/models/user-work.ts [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+ */
+
+/**
+ * Get a count of work items in the ready or running state for the given service ID
+ *
+ * @param tx - The database transaction
+ * @param serviceID - The ID of the service
+ * @returns The sum of ready and running work items for the service
+ */
+export async function getQueuedAndRunningCountForService(tx: Transaction, serviceID: string)
```
[the remark this answers is not in the document]
Comment (chris-durbin) [reply app/models/user-work.ts]: Done

File: app/models/user-work.ts [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+
+  // true if the requested job is asynchronous
+  is_async: boolean;
+
+  // the time the job was last worked
+  last_worked: Date;
+}
+
+/**
+ * Get user work record for jobID and serviceID
```
Comment (vinnyinverso) [inline app/models/user-work.ts:36]: Can this docstring be deleted?

File: app/models/services/base-service.ts [hunk 5]
```diff
@@ -412,6 +413,34 @@ export default abstract class BaseService<ServiceParamType> {
     return workflowSteps;
   }
 
+  /**
+   * Creates the workflow steps objects for this request
```
[the remark this answers is not in the document]
Comment (chris-durbin) [reply app/models/services/base-service.ts:417]: Fixed

File: app/models/user-work.ts (hunk 4 above)
Comment (chris-durbin) [reply app/models/user-work.ts:36]: Fixed

**Pull request merged**